src/orion/core/worker/warm_start/experiment_config.py

```diff
@@ -0,0 +1,127 @@
+from dataclasses import dataclass, field
+from typing import (
+    Any,
+    Dict,
+    List,
+    Optional,
+    Union,
+)
+from datetime import datetime
+from logging import getLogger as get_logger
+from .serializable import SerializableMixin
+
+logger = get_logger(__name__)
+
+
+@dataclass(frozen=True, unsafe_hash=True)
+class Config(SerializableMixin):
+    """Base class for the various Config objects: StorageConfig, ExperimentConfig, etc."""
+
+
+@dataclass(frozen=True)
+class DatabaseConfig(Config):
+    type: str
+
+
+@dataclass(frozen=True)
+class StorageConfig(Config):
+    type: str
+    database: DatabaseConfig
+
+
+@dataclass(frozen=True)
+class Refers(Config):
+    parent_id: Optional[str]
+    root_id: int
+    adapter: List[Any]
+
+
+@dataclass(frozen=True)
+class MetaData(Config):
+    user: str
+    """ System user currently owning this running process, the one who invoked **Oríon**. """
+
+    orion_version: str
+    """ Version of **Oríon** which suggested this experiment. `user`'s current **Oríon** version.
+    """
+
+    datetime: datetime
+    """ When was this particular configuration submitted to the database. """
+
+    # NOTE: Fields that don't appear to be set in the metadata, but that are mentioned in the
+    # docstring:
+
+    # user_script : str
+    #    Full absolute path to `user`'s executable.
+    # user_args : list of str
+    #    Contains separate arguments to be passed when invoking `user_script`,
+    #    possibly templated for **Oríon**.
+    # user_vcs : str, optional
+    #    User's version control system for this executable's code repository.
+    # user_version : str, optional
+    #    Current user's repository version.
+    # user_commit_hash : str, optional
+    #    Current `Experiment`'s commit hash for **Oríon**'s invocation.
+
+
+@dataclass(frozen=True)
+class ExperimentInfo(Config):
```

Comment:
This could be the start of a nice refactoring for Experiment and ExperimentClient. ExperimentInfo could clearly be useful outside of warm-start. Not suggesting for this PR though, but could be described in an issue for future work.

Reply:
I was thinking of more substantial changes in Experiment and ExperimentClient. Let's leave it here for now, but it would be good to move this when we are able to tackle a refactoring of Experiment and ExperimentClient.

Reply:
I moved this to `orion/core/worker/experiment_config.py` and also removed the Serializable class in 1b29a3813db57aa3f12e09cc49ca335fb5b15f4c

The `configuration` property is marked as returning this type of dictionary, although the implementation doesn't currently match (as far as I can tell). Therefore, I'm just leaving it as a type annotation, and adding a TODO/issue for later, to actually return the literal dictionary, with the right type of entries.

LMK what you think